src/java.base/share/classes/sun/security/util/Cache.java

```diff
@@ -591,4 +671,109 @@ public void invalidate() {
         }
     }
 
-}
+    /**
+     * This CacheEntry<K,V> type allows multiple V entries to be stored in
+     * one key in the cacheMap.
+     *
+     * This implementation is need for TLS clients that receive multiple
+     * PSKs or NewSessionTickets for server resumption.
+     */
+    private static class QueueCacheEntry<K,V> extends SoftReference<V>
+        implements CacheEntry<K,V> {
+
+        // Limit the number of queue entries.
+        private final int MAXQUEUESIZE;
+
+        final boolean DEBUG = false;
+        private K key;
+        private long expirationTime;
+        final Queue<CacheEntry<K,V>> queue = new ConcurrentLinkedQueue<>();
+
+        QueueCacheEntry(K key, V value, long expirationTime, int maxSize,
+            ReferenceQueue<V> queue) {
+            super(value, queue);
+            this.key = key;
+            this.expirationTime = expirationTime;
+            this.MAXQUEUESIZE = maxSize;
+        }
+
+        public K getKey() {
+            return key;
+        }
+
+        public V getValue() {
+            return getValue(0);
+        }
+
+        public V getValue(long lifetime) {
+            long time = (lifetime == 0) ? 0 : System.currentTimeMillis();
+            do {
+                var entry = queue.poll();
+                if (entry == null) {
+                    return null;
+                }
+                if (entry.isValid(time)) {
+                    // SoftReference get() returns the same as entry.getValue()
```

Comment:
this doesn't look right

Reply:
I checked it through debugging.  I was a bit surprised myself that it didn't return the `QueueCacheEntry`.get().  I can switch it the below if that's agreeable.
```
                if (entry.isValid(time)) {
                    if (entry instanceof SoftCacheEntry<K,V> sce) {
                        return sce.get();
                    }
                    return entry.getValue();
                }
```

Reply:
What what's it's worth, I changed the behavior here.  The SoftCacheEntry is now  only soft referenced, the QueueCacheEntry is not.